r-package/policytree/R/hybrid_policy_tree.R

```diff
@@ -0,0 +1,202 @@
+#' with `depth = 3` and `search.depth = 2`, the first split is determined by a depth 2 exact tree,
+#' and two new depth 2 trees are fit in the immediate children, leading to a total depth of 3.
+#' Note that the resulting tree may be shallower than the specified `depth` depending on
+#' whether leaf nodes were pruned or not.
+#' This algorithm scales with some coefficient multiple of the runtime of a `search.depth` `policy_tree`.
+#'
+#' The algorithm: desired depth is given by `depth`. Each node is split using exact tree search
+#' with depth  = `search.depth`. When we reach a node where the current level + `search.depth` is equal to `depth`,
+#' we stop and attach the `search.depth` subtree to this node.
+#' We also stop if the best `search.depth` split yielded a leaf node.
+#'
+#' @param X The covariates used. Dimension \eqn{N*p} where \eqn{p} is the number of features.
+#' @param Gamma The rewards for each action. Dimension \eqn{N*d} where \eqn{d} is the number of actions.
+#' @param depth The depth of the fitted tree. Default is 3.
+#' @param search.depth Depth to look ahead when splitting. Default is 2.
+#' @param split.step An optional approximation parameter, the number of possible splits
+#'  to consider when performing tree search. split.step = 1 (default) considers every possible split, split.step = 10
+#'  considers splitting at every 10'th sample and may yield a substantial speedup for dense features.
+#'  Manually rounding or re-encoding continuous covariates with very high cardinality in a
+#'  problem specific manner allows for finer-grained control of the accuracy/runtime tradeoff and may in some cases
+#'  be the preferred approach.
+#' @param min.node.size An integer indicating the smallest terminal node size permitted. Default is 1.
+#' @param verbose Give verbose output. Default is TRUE.
+#'
+#' @return A policy_tree object.
+#'
+#' @examples
+#' \donttest{
+#' # Fit a depth three tree on doubly robust treatment effect estimates from a causal forest.
+#' n <- 1500
+#' p <- 5
+#' X <- round(matrix(rnorm(n * p), n, p), 2)
+#' W <- rbinom(n, 1, 1 / (1 + exp(X[, 3])))
+#' tau <- 1 / (1 + exp((X[, 1] + X[, 2]) / 2)) - 0.5
+#' Y <- X[, 3] + W * tau + rnorm(n)
+#' c.forest <- grf::causal_forest(X, Y, W)
+#' dr.scores <- double_robust_scores(c.forest)
+#'
+#' tree <- hybrid_policy_tree(X, dr.scores, depth = 3)
+#'
+#' # Predict treatment assignment.
+#' predicted <- predict(tree, X)
+#' }
+#' @export
+hybrid_policy_tree <- function(X, Gamma,
+                               depth = 3,
+                               search.depth = 2,
+                               split.step = 1,
+                               min.node.size = 1,
+                               verbose = TRUE) {
+  if (search.depth >= depth) {
+    stop("`search.depth` should be less than `depth`.")
+  }
+  if (depth > 20) {
+    stop("Specified `depth` is too close to R's memory limit of 2^32 - 1.")
```

Comment:
You can also say that this is the exact error message R gives. otherwise, this can be confusing

Reply:
How about:

Reply:
How about:

"Specified depth is too large (as internal tree array has `2^(depth + 1) - 1` number of rows)."

Reply:
It is still not clear from the message what is going on. Maybe just say Specified depth is too large for number of nodes to fit in R's memory (because the limit should be 2^32 but it is 2^(21) - 1 here which is confusing me. On the whole, it is not too important, mostly pedantic.